Limit alarm handler walltime to 15minutes

jqmmes opened the pull request:
This PR adds a wall time timeout for alarm handlers. 

When the timeout for an alarm it hit, we cancel deferred alarm deletion and abort the currently running handler. 

The alarm timeouts are counted against limits, so they get deleted once the maximum retry limit is reached.

jclee commented on src/workerd/api/global-scope.c++:
```diff
@@ -386,21 +387,42 @@ kj::Promise<WorkerInterface::AlarmResult> ServiceWorkerGlobalScope::runAlarm(
+        // We don't want the handler to keep running after timeout.
+        context.abort(JSG_KJ_EXCEPTION(FAILED, Error, "Alarm exceeded its allowed execution time"));
+        // For now, we don't want to make the timeout destructive. As such, we will mark it as
+        // retriable, but let's now count retries against the limit. This behaviour will most likely
```
s/now/not ? (since it looks like this sets retryCountsAgainstLimit to false)

jclee commented on src/workerd/api/global-scope.c++:
```diff
@@ -386,21 +387,42 @@ kj::Promise<WorkerInterface::AlarmResult> ServiceWorkerGlobalScope::runAlarm(
+        return WorkerInterface::AlarmResult {
+          .retry = true,
+          .retryCountsAgainstLimit = false,
+          .outcome = EventOutcome::EXCEPTION
```
FWIW, it looks like the status for exceeding the `scheduledTimeoutMs` is `EventOutcome::EXCEEDED_CPU`?

jqmmes replied:
I'm not sure that we want to treat this error as `EventOutcome::EXCEEDED_CPU`. Ideally we'd have a specific reason for this, but since cron uses `EXCEEDED_CPU` I guess we can go with that for now.

dom96 commented on src/workerd/api/global-scope.c++:
```diff
@@ -386,21 +387,42 @@ kj::Promise<WorkerInterface::AlarmResult> ServiceWorkerGlobalScope::runAlarm(
+      // timeout will be canceled.
+      auto timeoutPromise = context.afterLimitTimeout(timeoutMs*kj::MILLISECONDS).then([&context]() -> kj::Promise<WorkerInterface::AlarmResult> {
+        LOG_NOSENTRY(WARNING, "Alarm exceeded its allowed execution time");
+        // We don't want to delete the alarm sicne we have not successfully completed the alarm
```
```suggestion
        // We don't want to delete the alarm since we have not successfully completed the alarm
```

dom96 commented on src/workerd/io/worker-entrypoint.c++:
```diff
@@ -551,11 +551,17 @@ kj::Promise<WorkerInterface::AlarmResult> WorkerEntrypoint::runAlarmImpl(
 
+          // If we have an invalid timeout, set it to the default value of 15 minutes.
+          auto timeoutMs = timeout;
+          if (timeout == 0) {
```
should we log an error or at least a warning in this case?

dom96 reviewed (approved):
Looks like this is tested upstream, nice.

How risky is this change? Would it make sense to put it behind an autogate to roll it out independent of a release?

Otherwise LGTM

jqmmes commented:
Updated PR with new expected behaviour where alarms will be deleted after 1 run + 6 retries.
- Updated description
- Added additional logging
- Alarm retries will now count against limits.

jqmmes commented:
Updated PR with timeout as `kj::Duration` instead of uint32_t.

justin-mp commented on src/workerd/api/global-scope.c++:
```diff
@@ -364,6 +364,7 @@ void ServiceWorkerGlobalScope::startScheduled(
 
 kj::Promise<WorkerInterface::AlarmResult> ServiceWorkerGlobalScope::runAlarm(
     kj::Date scheduledTime,
+    kj::Duration timeoutMs,
```
You'll probably hate me for saying this, but if you have a `kj::Duration`, the name should be `timeout` since the Duration handles units for you.  Please fix!

justin-mp commented on src/workerd/io/limit-enforcer.h:
```diff
@@ -120,6 +120,9 @@ class LimitEnforcer  {
   virtual kj::Promise<void> limitScheduled() = 0;
 
+  // Like limitDrain() and limitScheduled() but applies a time limit to alarm event processing.
+  virtual kj::Duration getAlarmLimitMs() = 0;
```
Similarly, since this returns a Duration, the name should be getAlarmLimit().

justin-mp commented on src/workerd/io/worker-entrypoint.c++:
```diff
@@ -551,11 +551,18 @@ kj::Promise<WorkerInterface::AlarmResult> WorkerEntrypoint::runAlarmImpl(
           jsg::AsyncContextFrame::StorageScope traceScope = context.makeAsyncTraceScope(lock);
 
+          // If we have an invalid timeout, set it to the default value of 15 minutes.
+          auto timeoutMs = timeout;
```
newTimeout?  Or maybe call the timeout from the enforcer `enforcerLimit`?

justin-mp commented on src/workerd/io/worker-entrypoint.c++:
```diff
@@ -551,11 +551,18 @@ kj::Promise<WorkerInterface::AlarmResult> WorkerEntrypoint::runAlarmImpl(
+          auto timeoutMs = timeout;
+          if (timeout == 0 * kj::MILLISECONDS) {
+            LOG_NOSENTRY(WARNING, "Invalid alarm timeout value. Using 900000ms", timeout);
+            timeoutMs = 900000 * kj::MILLISECONDS;
```
What's nice about using a duration is that you can set it to `15 * kj::MINUTES`, which is a whole lot easier to read.  (You may want to update the log message too.)

jqmmes commented:
Updated PR:
- Addressed @justin-mp comments
- Rebased
